File: leetcode-solutions/ruby/19. Remove Nth Node From End of List.rb
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
  slow, fast = head, head
  
  n.times do
      fast = fast.next
  end
  
  prev = slow
  while fast != nil
      prev = slow
      slow = slow.next
      fast = fast.next
  end
  
  if slow == head
      head = head.next
      return head
  else
      # puts slow.inspect
      prev.next = slow.nil? ? nil : slow.next
      head
  end
end

File: leetcode-solutions/python/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = slow = head
        # move fast n times, so that it reached to last node when slow will be before target node
        while n > 0:
            fast = fast.next
            n -= 1

        # have to remove head
        if fast is None:
            return head.next

        # untill find the last node so that slow point to target's previous node
        while fast.next:
            slow = slow.next
            fast = fast.next

        slow.next = slow.next.next
        return head
